File: strategies/1-brute-force/StringPermutations.java
import java.util.*;

/**
 * Brute Force Strategy: String Permutations
 * Core Idea: Generate all possible permutations by trying every combination
 * Time Complexity: O(n! * n) where n is the length of string
 * Space Complexity: O(n! * n) for storing all permutations
 */
public class StringPermutations {
    
    /**
     * Generate all permutations of a string using brute force approach
     * @param str Input string
     * @return List of all permutations
     */
    public static List<String> generatePermutations(String str) {
        List<String> result = new ArrayList<>();
        generatePermutationsHelper(str.toCharArray(), 0, result);
        return result;
    }
    
    /**
     * Helper method to generate permutations recursively
     * @param chars Character array to permute
     * @param index Current position being fixed
     * @param result List to store all permutations
     */
    private static void generatePermutationsHelper(char[] chars, int index, List<String> result) {
        // Base case: if we've fixed all positions, add the permutation
        if (index == chars.length) {
            result.add(new String(chars));
            return;
        }
        
        // Try each character at the current position
        for (int i = index; i < chars.length; i++) {
            // Swap current character with the character at index
            swap(chars, index, i);
            
            // Recursively generate permutations for remaining positions
            generatePermutationsHelper(chars, index + 1, result);
            
            // Backtrack: restore original order
            swap(chars, index, i);
        }
    }
    
    /**
     * Swap two characters in an array
     */
    private static void swap(char[] chars, int i, int j) {
        char temp = chars[i];
        chars[i] = chars[j];
        chars[j] = temp;
    }
    
    public static void main(String[] args) {
        // Test Case 1: Simple 3-character string
        System.out.println("=== String Permutations - Brute Force ===");
        String test1 = "ABC";
        System.out.println("Input: \"" + test1 + "\"");
        List<String> permutations1 = generatePermutations(test1);
        System.out.println("Output: " + permutations1);
        System.out.println("Total permutations: " + permutations1.size());
        System.out.println();
        
        // Test Case 2: String with repeated characters
        String test2 = "AAB";
        System.out.println("Input: \"" + test2 + "\"");
        List<String> permutations2 = generatePermutations(test2);
        System.out.println("Output: " + permutations2);
        System.out.println("Total permutations: " + permutations2.size());
        System.out.println();
        
        // Test Case 3: Single character
        String test3 = "X";
        System.out.println("Input: \"" + test3 + "\"");
        List<String> permutations3 = generatePermutations(test3);
        System.out.println("Output: " + permutations3);
        System.out.println("Total permutations: " + permutations3.size());
        System.out.println();
        
        // Test Case 4: Empty string
        String test4 = "";
        System.out.println("Input: \"" + test4 + "\"");
        List<String> permutations4 = generatePermutations(test4);
        System.out.println("Output: " + permutations4);
        System.out.println("Total permutations: " + permutations4.size());
    }
}

File: strategies/1-brute-force/StringPermutations.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

/**
 * Brute Force Strategy: String Permutations
 * Core Idea: Generate all possible permutations by trying every combination
 * Time Complexity: O(n! * n) where n is the length of string
 * Space Complexity: O(n! * n) for storing all permutations
 */

/**
 * Swap two characters in a string
 */
void swap(char* str, int i, int j) {
    char temp = str[i];
    str[i] = str[j];
    str[j] = temp;
}

/**
 * Generate all permutations of a string using brute force approach
 * @param str Input string
 * @param start Starting index for current permutation level
 * @param end Ending index (length - 1)
 */
void generatePermutations(char* str, int start, int end) {
    // Base case: if we've fixed all positions, print the permutation
    if (start == end) {
        printf("%s ", str);
        return;
    }
    
    // Try each character at the current position
    for (int i = start; i <= end; i++) {
        // Swap current character with the character at start
        swap(str, start, i);
        
        // Recursively generate permutations for remaining positions
        generatePermutations(str, start + 1, end);
        
        // Backtrack: restore original order
        swap(str, start, i);
    }
}

/**
 * Wrapper function to generate and print all permutations
 */
void printAllPermutations(char* str) {
    int len = strlen(str);
    if (len == 0) {
        printf("(empty string - 1 permutation)\n");
        return;
    }
    generatePermutations(str, 0, len - 1);
    printf("\n");
}

int main() {
    printf("=== String Permutations - Brute Force ===\n");
    
    // Test Case 1: Simple 3-character string
    char test1[] = "ABC";
    printf("Input: \"%s\"\n", test1);
    printf("Output: ");
    printAllPermutations(test1);
    printf("Total permutations: 6 (3!)\n\n");
    
    // Test Case 2: String with repeated characters
    char test2[] = "AAB";
    printf("Input: \"%s\"\n", test2);
    printf("Output: ");
    printAllPermutations(test2);
    printf("Total permutations: 6 (includes duplicates)\n\n");
    
    // Test Case 3: Single character
    char test3[] = "X";
    printf("Input: \"%s\"\n", test3);
    printf("Output: ");
    printAllPermutations(test3);
    printf("Total permutations: 1\n\n");
    
    // Test Case 4: Two characters
    char test4[] = "AB";
    printf("Input: \"%s\"\n", test4);
    printf("Output: ");
    printAllPermutations(test4);
    printf("Total permutations: 2 (2!)\n\n");
    
    // Test Case 5: Empty string
    char test5[] = "";
    printf("Input: \"%s\"\n", test5);
    printf("Output: ");
    printAllPermutations(test5);
    
    return 0;
}